### DISCUSSION TITLE
How to change the index type of gsSparseMatrix?

### DISCUSSION THREAD

**Discussion by Rennngw** [gismo/gismo#684] [Q&A]
Hi everyone, I recently got some problems when developing my code using gismo.

Recently I want to change the type of matrix index in gismo so that it can perform large-scale matrix operations. But I found that when I changed the index type of gsSparseMatrix from int to long long, I used the following code to construct a gsSparseMatrix object.

```
gsSparseMatrix<double, 0, long long> k(1000, 1000);
gsSparseEntries<double> entries;
for (int i = 0; i < 100; i++)
	entries.add(i, i, 1.0);
k.setFrom(entries);
```
After the initialization is completed, I use gsSparseEntries to fill the gsSparseMatrix, but when I use the setFrom command, an std::bad_alloc exception will be reported. But I have enough memory.
Then I only use the eigen library to run the following code that completes similar functions, it can run normally.

```
Eigen::SparseMatrix<double, 0, long long> mat(1000, 1000);
std::vector<Eigen::Triplet<double>> vec;
for (int i = 0; i < 100; i++) {
	vec.push_back(Eigen::Triplet<double>(i, i, 1.0));
}
mat.setFromTriplets(vec.begin(), vec.end());
```

Through debugging, I found that the definition of the macro EIGEN_DEFAULT_DENSE_INDEX_TYPE in the eigen library is
`#define EIGEN_DEFAULT_DENSE_INDEX_TYPE std::ptrdiff_t`

while in gismo the macro EIGEN_DEFAULT_DENSE_INDEX_TYPE is defined as
`#define EIGEN_DEFAULT_DENSE_INDEX_TYPE index_t`

I tried to change the macro EIGEN_DEFAULT_DENSE_INDEX_TYPE in gismo to std::ptrdiff_t. After the modification, the above code for filling the sparse matrix can run normally, but exceptions will be reported in other locations, such as degreeElevate. I would like to ask if I want to modify the index of gsSparseMatrix, what should I do? How is it done?

**Comment by hverhelst** [comment]
Hi @Rennngw , we usually compile gismo with the CMAKE option `GISMO_INDEX_TYPE=long int` if we want  larger sparse matrices. The `gsSparseEntries` takes this type for its indices as well.
Does this help you out?
